File: clients/swift/Sources/HhmClient/ServiceClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct ApiError: Error { public let status: Int; public let body: Any? }

public final class ServiceClient {
    private let baseURL: URL
    private let token: String?
    private let session: URLSession

    public init(baseURL: URL, token: String? = nil, session: URLSession = .shared) {
        self.baseURL = baseURL; self.token = token; self.session = session
    }
    public func health() async throws -> Any? { try await request(path: "/healthz") }
    public func ready() async throws -> Any? { try await request(path: "/readyz") }
    public func config() async throws -> Any? { try await request(path: "/api/config") }
    public func emitEvent(_ payload: Any) async throws -> Any? { try await request(path: "/api/events", method: "POST", body: payload) }
    public func createLead(_ payload: Any) async throws -> Any? { try await request(path: "/api/leads", method: "POST", body: payload) }
    public func createAlert(_ payload: Any) async throws -> Any? { try await request(path: "/api/alerts", method: "POST", body: payload) }
    public func request(path: String, method: String = "GET", body: Any? = nil) async throws -> Any? {
        let relative = path.hasPrefix("/") ? String(path.dropFirst()) : path
        var request = URLRequest(url: baseURL.appendingPathComponent(relative))
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "accept")
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "authorization") }
        if let body { request.setValue("application/json", forHTTPHeaderField: "content-type"); request.httpBody = try JSONSerialization.data(withJSONObject: body) }
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        let parsed = data.isEmpty ? nil : try JSONSerialization.jsonObject(with: data)
        guard (200..<300).contains(http.statusCode) else { throw ApiError(status: http.statusCode, body: parsed) }
        return parsed
    }
}

File: clients/php/src/ServiceClient.php
<?php

declare(strict_types=1);

namespace HackerHouseMedellin\Client;

use RuntimeException;

final class ApiException extends RuntimeException
{
    public function __construct(public readonly int $status, public readonly mixed $body)
    {
        parent::__construct("request failed: {$status}");
    }
}

final class ServiceClient
{
    public function __construct(private string $baseUrl, private ?string $token = null)
    {
        if ($baseUrl === '') throw new \InvalidArgumentException('baseUrl is required');
        $this->baseUrl = rtrim($baseUrl, '/');
    }
    public function health(): mixed { return $this->request('/healthz'); }
    public function ready(): mixed { return $this->request('/readyz'); }
    public function config(): mixed { return $this->request('/api/config'); }
    public function emitEvent(mixed $payload): mixed { return $this->request('/api/events', 'POST', $payload); }
    public function createLead(mixed $payload): mixed { return $this->request('/api/leads', 'POST', $payload); }
    public function createAlert(mixed $payload): mixed { return $this->request('/api/alerts', 'POST', $payload); }

    public function request(string $path, string $method = 'GET', mixed $body = null): mixed
    {
        $headers = ['accept: application/json'];
        if ($this->token !== null) $headers[] = "authorization: Bearer {$this->token}";
        $content = null;
        if ($body !== null) {
            $headers[] = 'content-type: application/json';
            $content = json_encode($body, JSON_THROW_ON_ERROR);
        }
        $context = stream_context_create(['http' => ['method' => $method, 'header' => implode("\r\n", $headers), 'content' => $content, 'ignore_errors' => true]]);
        $raw = file_get_contents($this->baseUrl . '/' . ltrim($path, '/'), false, $context);
        $status = 0;
        if (isset($http_response_header[0]) && preg_match('/\s(\d{3})\s/', $http_response_header[0], $match)) $status = (int) $match[1];
        $parsed = $raw === false || $raw === '' ? null : json_decode($raw, true, flags: JSON_THROW_ON_ERROR);
        if ($status < 200 || $status >= 300) throw new ApiException($status, $parsed);
        return $parsed;
    }
}
